src/types.h

```diff
@@ -42,13 +51,324 @@ struct NetplanOptionalAddressType {
+    NETPLAN_WIFI_WOWLAN_4WAY_HANDSHAKE    = 1<<6,
+    NETPLAN_WIFI_WOWLAN_RFKILL_RELEASE    = 1<<7,
+    NETPLAN_WIFI_WOWLAN_TCP               = 1<<8,
+} NetplanWifiWowlanFlag;
+
+struct NetplanWifiWowlanType {
+    char* name;
+    NetplanWifiWowlanFlag flag;
+};
+
 extern struct NetplanWifiWowlanType NETPLAN_WIFI_WOWLAN_TYPES[];
 
+typedef enum {
+    NETPLAN_AUTH_KEY_MANAGEMENT_NONE,
+    NETPLAN_AUTH_KEY_MANAGEMENT_WPA_PSK,
+    NETPLAN_AUTH_KEY_MANAGEMENT_WPA_EAP,
+    NETPLAN_AUTH_KEY_MANAGEMENT_8021X,
+    NETPLAN_AUTH_KEY_MANAGEMENT_MAX,
+} NetplanAuthKeyManagementType;
+
+typedef enum {
+    NETPLAN_AUTH_EAP_NONE,
+    NETPLAN_AUTH_EAP_TLS,
+    NETPLAN_AUTH_EAP_PEAP,
+    NETPLAN_AUTH_EAP_TTLS,
+    NETPLAN_AUTH_EAP_METHOD_MAX,
+} NetplanAuthEAPMethod;
+
+typedef enum
+{
+    /**
+     * @brief Tristate enum type
+     * 
+     * This type defines a boolean which can be unset, i.e.
+     * this type has three states. The enum is ordered so
+     * that
+     * 
+     * FALSE -> 0
+     * TRUE  -> 1
+     * UNSET -> 2
+     * 
+     * And the integer values can be used directly when
+     * converting to string.
+     */
+    NETPLAN_TRISTATE_FALSE, /* 0 */
+    NETPLAN_TRISTATE_TRUE,  /* 1 */
+    NETPLAN_TRISTATE_UNSET, /* 2 */
+} NetplanTristate;
+
 typedef struct missing_node {
     char* netdef_id;
     const yaml_node_t* node;
 } NetplanMissingNode;
 
+typedef struct authentication_settings {
+    NetplanAuthKeyManagementType key_management;
+    NetplanAuthEAPMethod eap_method;
+    char* identity;
+    char* anonymous_identity;
+    char* password;
+    char* ca_certificate;
+    char* client_certificate;
+    char* client_key;
+    char* client_key_password;
+    char* phase2_auth;  /* netplan-feature: auth-phase2 */
+} NetplanAuthenticationSettings;
+
+/* Fields below are valid for dhcp4 and dhcp6 unless otherwise noted. */
+typedef struct dhcp_overrides {
+    gboolean use_dns;
+    gboolean use_ntp;
+    gboolean send_hostname;
+    gboolean use_hostname;
+    gboolean use_mtu;
+    gboolean use_routes;
+    char* use_domains; /* netplan-feature: dhcp-use-domains */
+    char* hostname;
+    guint metric;
+} NetplanDHCPOverrides;
+
+typedef struct ovs_controller {
+    char* connection_mode;
+    GArray* addresses;
+} NetplanOVSController;
+
+typedef struct ovs_settings {
+    GHashTable* external_ids;
+    GHashTable* other_config;
+    char* lacp;
+    char* fail_mode;
+    gboolean mcast_snooping;
+    GArray* protocols;
+    gboolean rstp;
+    NetplanOVSController controller;
+    NetplanAuthenticationSettings ssl;
+} NetplanOVSSettings;
+
+typedef union {
+    struct NetplanNMSettings {
+        char *name;
+        char *uuid;
+        char *stable_id;
+        char *device;
+        GData* passthrough; /* See g_datalist* functions */
+    } nm;
+    struct NetplanNetworkdSettings {
+        char *unit;
+    } networkd;
+} NetplanBackendSettings;
+
+struct netplan_net_definition {
```

Comment:
We might have a merge conflict here, as many of the definitions from `types.h` actually moved into `abi.h` recently.

Reply:
I had rebased this update and moved some of the changes into `abi.h`. I hope it was all correct :smile:

Reply:
No, definitely not. Look at types.h in this PR to see what I mean.

Reply:
Thanks @rlaager and @slyon . I see what you meant now. That rebase definitely didn't go as planned :/